src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs

```diff
@@ -68,46 +69,142 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
-                        var underlyingField2 = (FieldSymbol) Binder.GetWellKnownTypeMember(binder.Compilation, WellKnownMember.System_Runtime_CompilerServices_ValueTuple_T1_T2__Item2, diagnostics, syntax: syntax);
-
-                        fields = ImmutableArray.Create(
-                                    new TupleFieldSymbol(elementNames.IsDefault ? "Item1" : elementNames[0],
-                                        this,
-                                        elementTypes[0],
-                                        underlyingField1?.AsMember(underlyingType)),
-                                    new TupleFieldSymbol(elementNames.IsDefault ? "Item2" : elementNames[1],
-                                        this,
-                                        elementTypes[1],
-                                        underlyingField2?.AsMember(underlyingType))
-                                );
-
-                        return underlyingType;
-                    }
+                var tupleType = binder.GetWellKnownType(GetTupleType(elementTypes.Length), diagnostics, syntax);
+                underlyingType = tupleType.Construct(elementTypes);
 
-                default:
-                    {
-                        // TODO: VS if this eventually still stays reachable, need to make some error type symbol
-                        var tupleType = binder.GetWellKnownType(WellKnownType.System_Runtime_CompilerServices_ValueTuple_T1_T2, diagnostics, syntax);
-                        underlyingType = tupleType.Construct(elementTypes);
-                        break;
-                    }
+                var fieldSymbols = ArrayBuilder<TupleFieldSymbol>.GetInstance(elementTypes.Length);
 
+                for (int i = 0; i < elementTypes.Length; i++)
+                {
+                    var wellKnownTypeMember = GetTupleTypeMember(elementTypes.Length, i);
+                    var underlyingField = (FieldSymbol)Binder.GetWellKnownTypeMember(binder.Compilation, wellKnownTypeMember, diagnostics, syntax: syntax);
+                    var field = new TupleFieldSymbol(elementNames.IsDefault ? UnderlyingMemberName(i) : elementNames[i],
+                                               this,
+                                               elementTypes[i],
+                                               underlyingField?.AsMember(underlyingType));
+                    fieldSymbols.Add(field);
+                }
+
+                fields = fieldSymbols.ToImmutableAndFree();
+            }
+            else
+            {
+                var tupleType = binder.GetWellKnownType(WellKnownType.System_Runtime_CompilerServices_ValueTuple_T1, diagnostics, syntax);
+                fields = ImmutableArray<TupleFieldSymbol>.Empty;
+                underlyingType = tupleType.Construct(ImmutableArray.Create(elementTypes, 0 , 1));
```

Comment:
I would suggest to make this code unreachable, by making changes to BindTupleExpression and BindTupleType to report an error and bail early, without even attempting to create a TupleTypeSymbol if number of items is out of supported range. Here, we would just throw unreachable.